### PR TITLE
feat(node): allow executing esm compiled scripts

### LINKED ISSUE #10565: Unable to execute node ES Modules
## Current Behavior
When attempting to execute an ES Module with the `@nrwl/node:node` executor, we get the error reported in the below [Failure Logs section](#failure-logs). After debugging, I've identified that the issue is because the executor is attempting to `require` the module, which is of course not supported for ES Modules.

## Expected Behavior
The ES Module node.js application should properly execute without errors.

## Steps to Reproduce
[Pull this repo](https://github.com/cadenzatv/nx-issue-es-module) in order to reproduce.

1. Create a new node.js project ` nx generate @nrwl/node:app` 
2. Update `tsconfig.app.json` with the following fields:
```json
"module": "esnext",
"target": "es2020"
```
3. Update `project.json` `build` target to use executor `@nrwl/node:webpack`
4. Update `project.json` `build` target with the following fields
```json
"outputFileName": "main.mjs",
"webpackConfig": "webpack.config.js"
```
5. Update `project.json` `serve` target to use executor `@nrwl/node:node`
6. Create root `webpack.config.js` file with the following content
```javascript
module.exports = (config, context) => ({
    ...config,
    experiments: {
        ...config.experiments,
        outputModule: true,
        topLevelAwait: true,
    },
    output: {
        path: config.output.path,
        chunkFormat: 'module',
        library: {
            type: 'module',
        },
    },
})
```

### Failure Logs
```
/app/node_modules/@nrwl/node/src/executors/node/node-with-require-overrides.js:16
        return originalLoader.apply(this, arguments);
                              ^
Error [ERR_REQUIRE_ESM]: require() of ES Module /app/dist/packages/test/main.mjs not supported.
Instead change the require of /app/dist/packages/test/main.mjs to a dynamic import() which is available in all CommonJS modules.
    at Function.Module._load (/app/node_modules/@nrwl/node/src/executors/node/node-with-require-overrides.js:16:31)
    at Object.<anonymous> (/app/node_modules/@nrwl/node/src/executors/node/node-with-require-overrides.js:19:1)
```

### Environment
```
$ nx report

 >  NX   Report complete - copy this into the issue template

   Node : 16.13.0
   OS   : darwin x64
   npm  : 8.1.0
   
   nx : 14.1.7
   @nrwl/angular : Not Found
   @nrwl/cypress : Not Found
   @nrwl/detox : Not Found
   @nrwl/devkit : 14.1.7
   @nrwl/eslint-plugin-nx : 14.1.7
   @nrwl/express : Not Found
   @nrwl/jest : 14.1.7
   @nrwl/js : 14.1.7
   @nrwl/linter : 14.1.7
   @nrwl/nest : 14.1.7
   @nrwl/next : Not Found
   @nrwl/node : 14.1.7
   @nrwl/nx-cloud : 14.0.3
   @nrwl/nx-plugin : Not Found
   @nrwl/react : Not Found
   @nrwl/react-native : Not Found
   @nrwl/schematics : Not Found
   @nrwl/storybook : Not Found
   @nrwl/web : Not Found
   @nrwl/workspace : 14.1.7
   typescript : 4.3.5
   rxjs : 7.2.0
   ---------------------------------------
   Community plugins:

```

### LINKED ISSUE #16036: Error: Dynamic require of "*" is not supported
### Current Behavior

When I build my node app as esm module, it unable to run with node, because of following error:

```log
file:///Users/whalemare/ts-pipeline/dist/apps/node-example/main.js:12
  throw new Error('Dynamic require of "' + x + '" is not supported');
```

### Expected Behavior

My `node-example` app builded with success

### GitHub Repo

https://github.com/whalemare/ts-pipeline

### Steps to Reproduce

1. Clone repo and install dependencies

```
git clone https://github.com/whalemare/ts-pipeline
cd ts-pipeline
git checkout feature/esbuild
yarn install
```

2. Build and run example
```
nx run node-example:build
node dist/apps/node-example/main.js
```

### Nx Report

```shell
λ  ts-pipeline feature/esbuild ✓ yarn nx report
yarn run v1.22.17
$ /Users/whalemare/ts-pipeline/node_modules/.bin/nx report
>  NX  Falling back to ts-node for local typescript execution. This may be a little slower.
  - To fix this, ensure @swc-node/register and @swc/core have been installed

 >  NX   Report complete - copy this into the issue template

   Node : 18.15.0
   OS   : darwin arm64
   yarn : 1.22.17
   
   nx                      : 15.9.2
   @nrwl/js                : 15.9.2
   @nrwl/jest              : 15.9.2
   @nrwl/linter            : 15.9.2
   @nrwl/workspace         : 15.9.2
   @nrwl/cli               : 15.9.2
   @nrwl/cypress           : 15.9.2
   @nrwl/devkit            : 15.9.2
   @nrwl/esbuild           : 15.9.2
   @nrwl/eslint-plugin-nx  : 15.9.2
   @nrwl/node              : 15.9.2
   @nrwl/react             : 15.9.2
   @nrwl/tao               : 15.9.2
   @nrwl/vite              : 15.9.2
   @nrwl/web               : 15.9.2
   @nrwl/webpack           : 15.9.2
   typescript              : 4.9.4
```

### Failure Logs

```shell
file:///Users/whalemare/ts-pipeline/dist/apps/node-example/main.js:12
  throw new Error('Dynamic require of "' + x + '" is not supported');
```

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by bulldog98**
## Current Behavior
The @nrwl/node:node executor is not able to execute esm module scripts

## Expected Behavior
The @nrwl/node:node executor is able to execute esm module scripts

Fixes: #10565
Fixes: #16036

### CODE REVIEW

**Comment by bulldog98** [comment]
Why did the build not run?

**Comment by nartc** [comment]
@bulldog98 , thank you for the PR. Can you add some e2e tests for this change?

**Comment by bulldog98** [comment]
> @bulldog98 , thank you for the PR. Can you add some e2e tests for this change?
Sure I first did not see there where e2e tests, but found them now.

**Comment by bulldog98** [comment]
@nartc I was unable to run the e2e test locally, how are they executed?

**Comment by nartc** [comment]
@bulldog98 I've rebased the branch and kicked off the build which is failing. Can you take a look? To run test locally, you'd need to start `local-registry` with:

```
yarn local-registry enable
yarn local-registry start
```

then run the e2e:

```
yarn nx e2e e2e-node
```

**Comment by bulldog98** [comment]
sure I'm working on fixing the test.

**Comment by bulldog98** [comment]
@nartc I fixed the e2e test it works now

**Comment by StringKe** [comment]
I had the same problem.
How can I use this pr to test if it's the same problem?

**Comment by bulldog98** [comment]
@nartc sorry I had messed up the check to beEqual, but it should be a toContains

**Comment by nartc** [comment]
@bulldog98 Thanks for the PR. Everything looks good so far. Just one quick question, why did you have the following?

```ts
...
const dynamicImport = new Function('specifier', 'return import(specifier)');
...
```

Would this work as well?

```ts
...
function dynamicImport(specifier) {
   return import(specifier);
}

dynamicImport(fileToRun);
...
```

**Comment by luxaritas** [comment]
That said, I'm curious why not just `import(fileToRun)`

**Comment by bulldog98** [comment]
The problem is that typescript transpiles the `import(fileToRun)` to `require(fileToRun)` the same happens with the version of @nartc.

**Comment by luxaritas** [comment]
The "real" fix would be to change "module" to "es2020" in the tsconfig - that would be a much larger change, but also this implicitly breaks environments that don't support the dynamic import syntax natively instead of explicitly. I think all current supported node versions should though?

**Comment by nartc** [comment]
@luxaritas yep, that's my concern as well. I also think the real fix is to support ESM across the board. With TS 4.7, we have a more "official" ESM support but we're somewhat waiting for the dust to settle before making substantial changes to all plugins.

**Comment by LinboLen** [comment]
nrwl build app support esm will fix some problem. use custom webpack 

project.json
```json5
{
   "targets": {
     "build": {
      "executor": "@nrwl/node:webpack",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/apps/<project>",
        "main": "apps/<project>/src/main.ts",
        "tsConfig": "apps/<project>/tsconfig.app.json",
        "assets": ["apps/<project>/src/assets"],
        "webpackConfig": "apps/<project>/webpack.config.js"   <== add this
      },
  }
}
```


apps/<project>/webpack.config.js
```javascript
module.exports = (config, context) => ({
  ...config,
  experiments: {
    ...config.experiments,
    outputModule: true,
    topLevelAwait: true,
  },
  output: {
    path: config.output.path,
    chunkFormat: 'module',
    library: {
      type: 'module',
    },
  },
})

```

**Comment by nartc** [comment]
Closing because we'd like to implement a better fix across Nx to support ESM instead of dynamic import hack. Thank you bulldog98 for the PR but I'll be closing this.

**Comment by kevinkmp** [comment]
@nartc - This problem has completely derailed one of our active projects and it seems like many others on this thread are struggling with this also. Because of that I truly hope that closing this PR means that there will be an official fix released in the **_very_** near future.

**Comment by qwertyuiopzxcvbnmlkjhgfdsa1** [comment]
Created similar bug to this one  https://github.com/nrwl/nx/issues/12259

**Comment by sndrs** [comment]
@nartc that's great news! is there anywhere we can track/contribute to this?

**Comment by omsharp** [comment]
I hope this issue is solved soon!
I've just upgraded my project to v14 and now I'm not able to do anything with it unless this issue is solved! My project is practically frozen!

**Comment by antonin-san** [comment]
why is it so complicated, seriously, i just want to set my f*cking desk on fire. 3 hours trying to add an app to NX that only uses yargs and inquirer :face_exhaling : , i quit

**Comment by StringKe** [comment]
> why is it so complicated, seriously, i just want to set my f*cking desk on fire. 3 hours trying to add an app to NX that only uses yargs and inquirer 😮‍💨 , i quit

try this https://github.com/esbuild-kit/tsx just run command

**Comment by goldylucks** [comment]
as a workaround, is it possible to tell webpack to handle a specific node module as esm?

psuedo code:

```
// webpack.config.js
module: {
      rules: [
        {
          test: /node_modules/the-package,
          target: "module",
loader: "babel-loader",
        },
      ],
    },
```

**Comment by codeedog** [comment]
@nartc 15.3.0 came out today, yet this is still broken. What's up?

**Comment by quaelin** [comment]
I just upgraded to 15.4.2, and my project is borked.  I'm not even sure there's a possible path forward until nx has ESM support... but does anyone have a workaround in the meantime?

**Comment by bulldog98** [comment]
@quaelin you can patch the file in the error directly and replace the require with import

**Comment by StringKe** [comment]
@quaelin 

I am using the following solution to solve the esm problem, you can refer to use it.

```javascript
// apps/SOME/webpack.config.cjs

const nodeExternals = require('webpack-node-externals')

module.exports = (config, context) => {
	return {
		...config,
		externalsPresets: {
			node: true,
		},
		output: {
			...config.output,
			module: true,
			libraryTarget: 'module',
			chunkFormat: 'module',
			library: {
				type: 'module',
			},
			environment: {
				module: true,
			},
		},
		experiments: {
			outputModule: true,
		},
		externals: nodeExternals({
			importType: 'module',
		}),
	}
}

```

```json
// project.json or workspace.json
                    "build": {
			"executor": "@nrwl/webpack:webpack",
			"dependsOn": ["build-porto"],
			"outputs": ["{options.outputPath}"],
			"options": {
				"target": "node",
				"compiler": "tsc",
				"outputPath": "dist/apps/api",
				"main": "apps/api/src/main.ts",
				"tsConfig": "apps/api/tsconfig.app.json",
				"assets": ["apps/api/src/assets"],
                                 // add next line
				"webpackConfig": "apps/api/webpack.config.cjs",
				"generatePackageJson": true
			},
```

add package.json or create package.json
```
//apps/SOME/package.json

"type": "module"
```

update project tsconfig.json
```json
// tsconfig.app.json
{
	"extends": "./tsconfig.json",
	"compilerOptions": {
		"outDir": "../../dist/out-tsc",
                 // change this
		"module": "es2020",
		"types": ["node"]
	},
	"exclude": ["jest.config.ts", "src/**/*.spec.ts", "src/**/*.test.ts"],
	"include": ["src/**/*.ts"]
}

```

### pnpm patch

1. pnpm patch @nrwl/js
2. modify `src/executors/node/node-with-require-overrides.js`
```diff
-require(fileToRun);
+const dynamicImport = new Function('specifier', 'return import(specifier)');
+dynamicImport(fileToRun);
```
3. pnpm patch-commit [path], path from step 1

In addition to using pnpm patch you can also use https://github.com/ds300/patch-package

**Comment by tuyen-vuduc** [comment]
@nartc @luxaritas Any plan for the official fix to be rolled out?

**Comment by gaurang171** [comment]
does any one knows what was the last working version? may be its better to downgrade

**Review by jaysoo — APPROVED** [review]

**Comment by jaysoo** [comment]
We'll merge this fix, and introduce even more functionality to make sure CJS/ESM compatibility is consistent between the different technologies (webpack, esbuild, tsc, swc, etc.). This is a good workaround until then.

File: lerna.json
Comment (AgentEnder) [inline lerna.json:1]: This probably slipped in 😄

**Pull request merged**